mlflow/pyspark/ml/__init__.py

```diff
@@ -107,24 +120,129 @@ def _get_estimator_info_tags(estimator):
     }
 
 
-def _get_pipeline_stage_hierarchy(pipeline):
+def _is_parameter_search_estimator(instance):
+    from pyspark.ml.tuning import CrossValidator, TrainValidationSplit
+    return isinstance(instance, (CrossValidator, TrainValidationSplit))
+
+
+def _is_parameter_search_model(instance):
+    from pyspark.ml.tuning import CrossValidatorModel, TrainValidationSplitModel
+    return isinstance(instance, (CrossValidatorModel, TrainValidationSplitModel))
+
+
+def _should_log_hierarchy(estimator):
+    from pyspark.ml import Pipeline
+    from pyspark.ml.classification import OneVsRest
+    return isinstance(estimator, (Pipeline, OneVsRest)) or \
+        _is_parameter_search_estimator(estimator)
+
+
+AutologgingEstimatorMetadata = namedtuple('AutologgingEstimatorMetadata', [
+    'hierarchy', 'uid2name_map', 'param_search_estimators'
+])
+
+
+def _gen_instance_metadata_recursively(
+        instance,
+        uid2name_map,
+        cls2count_map,
+        cls2first_uid_map,
+        param_search_estimators,
+):
```

Comment:
Can we add some docs to this method including docstrings for the parameters? From reading the code, the meaning of these `cls2...` maps is not so obvious